pkg/controller/cloneset/update/cloneset_update.go

```diff
@@ -1,5 +1,256 @@
+}
+
+func (c *realControl) Manage(cs *appsv1alpha1.CloneSet,
+	updateRevision *apps.ControllerRevision, revisions []*apps.ControllerRevision,
+	pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim,
+) error {
+
+	if cs.Spec.UpdateStrategy.Paused {
+		return nil
+	}
+
+	// 1. find currently updated and not-ready count and all pods waiting to update
+	var updatedNotReadyCount int
+	var waitUpdateIndexes []int
+	for i := range pods {
+		if inplaceupdate.IsToUpdateConditionReady(pods[i]) {
+			klog.V(4).Infof("StatefulSet %s/%s update Pod %s InPlaceUpdateReady condition to true", cs.Namespace, cs.Name, pods[i].Name)
+			newCondition := v1.PodCondition{
+				Type:               appsv1alpha1.InPlaceUpdateReady,
+				Status:             v1.ConditionTrue,
+				LastTransitionTime: c.now(),
+			}
+			var err error
+			if pods[i], err = c.updatePodCondition(pods[i], newCondition); err != nil {
+				return err
+			}
+		}
+
+		if clonesetutils.GetPodRevision(pods[i]) != updateRevision.Name {
+			waitUpdateIndexes = append(waitUpdateIndexes, i)
+		} else if !readyForUpdate(pods[i]) {
+			updatedNotReadyCount++
+		}
+	}
+
+	// 2. sort all pods waiting to update
+	waitUpdateIndexes = sortUpdateIndexes(cs.Spec.UpdateStrategy, pods, waitUpdateIndexes)
+
+	// 3. calculate max count of pods can update
+	needToUpdateCount := calculateUpdateCount(cs.Spec.UpdateStrategy, int(*cs.Spec.Replicas), len(waitUpdateIndexes), updatedNotReadyCount)
+	if needToUpdateCount < len(waitUpdateIndexes) {
+		waitUpdateIndexes = waitUpdateIndexes[:needToUpdateCount]
+	}
+
+	// 4. update pods
+	for _, idx := range waitUpdateIndexes {
+		pod := pods[idx]
+		if err := c.updatePod(cs, updateRevision, revisions, pod, pvcs); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func sortUpdateIndexes(strategy appsv1alpha1.CloneSetUpdateStrategy, pods []*v1.Pod, waitUpdateIndexes []int) []int {
+	// not-ready < ready, unscheduled < scheduled, and pending < running
+	sort.Slice(waitUpdateIndexes, func(i, j int) bool {
+		return kubecontroller.ActivePods(pods).Less(waitUpdateIndexes[i], waitUpdateIndexes[j])
+	})
+	if strategy.PriorityStrategy != nil {
+		// TODO(FillZpp): sort by priority
+	}
+	return waitUpdateIndexes
+}
+
+func calculateUpdateCount(strategy appsv1alpha1.CloneSetUpdateStrategy, totalReplicas, notUpdatedCount, updatedNotReadyCount int) int {
+	partition := 0
+	if strategy.Partition != nil {
+		partition = int(*strategy.Partition)
+	}
+	maxUnavailable, _ := intstrutil.GetValueFromIntOrPercent(
+		intstrutil.ValueOrDefault(strategy.MaxUnavailable, intstrutil.FromString("10%")), totalReplicas, true)
+
+	return integer.IntMax(integer.IntMin(
+		notUpdatedCount-partition,
+		maxUnavailable-updatedNotReadyCount,
+	), 0)
+}
+
+func (c *realControl) updatePod(cs *appsv1alpha1.CloneSet,
+	updateRevision *apps.ControllerRevision, revisions []*apps.ControllerRevision,
+	pod *v1.Pod, pvcs []*v1.PersistentVolumeClaim,
+) error {
+	var inPlaceUpdateSpec *inplaceupdate.UpdateSpec
```

Comment:
Is the following block of code a dup of isToInPlaceUpdate method in stateful_set_util.go?

Reply:
Moved to ./pkg/util/inplaceupdate